### PR TITLE
Update broad phase to work properly with world ids

### PR DESCRIPTION
**Pull request by nvtw**
## Description

Should address https://github.com/newton-physics/newton/issues/946

This PR improves the broad phase collision detection system with efficiency optimizations and consistency fixes:

### Core Changes
- **Optional pair precomputation**: Added `build_shape_contact_pairs` parameter to `ModelBuilder.finalize()` to skip expensive O(N²) computation when using dynamic broad phases (NXN/SAP)
- **Shape flags filtering**: Extended `precompute_world_map()` to filter out non-colliding shapes (e.g., visual-only) from broad phase processing, preventing unnecessary work
- **Terminology fixes**: Clarified distinction between collision groups and their filtering behavior
- **Type fixes**: Converted `model.shape_collision_group` from Python list to warp array for kernel compatibility

### Enhancements
- **SAP sort configuration**: Added `sap_sort_type` parameter to control sorting algorithm (SEGMENTED vs TILE) in SAP broad phase
- **Consistent filtering**: All three broad phase modes (EXPLICIT, NXN, SAP) now consistently respect `ShapeFlags.COLLIDE_SHAPES`
- **Comprehensive testing**: Parametrized unified collision pipeline tests to run for all three broad phase modes (27 tests total)

## Newton Migration Guide

Please ensure the migration guide for **warp.sim** users is up-to-date with the changes made in this PR.

- [x] The migration guide in ``docs/migration.rst`` is up-to date

## Before your PR is "Ready for review"

- [x] Necessary tests have been added and new examples are tested (see `newton/tests/test_examples.py`)
- [x] Documentation is up-to-date
- [x] Code passes formatting and linting checks with `pre-commit run -a`

## Summary by CodeRabbit

* **New Features**
  * Public SAPSortType added; new options to supply per-shape world, flags, collision groups, and SAP sort selection.

* **Enhancements**
  * Multi-world broad-phase with per-world precomputation and world-aware pair generation for NXN and SAP.
  * Segmented (default) and optional tile-based SAP sorting for improved configurability and performance.

* **Tests**
  * Expanded coverage for multi-world, SAP segmented/tile variants, flags, and edge cases.

* **Chores**
  * shape_collision_group now finalized and stored as an array; model initialization updated.

### CODE REVIEW

File: newton/_src/geometry/broad_phase_sap.py [hunk 1]
```diff
@@ -97,98 +55,175 @@ def _sap_project_aabb(
...
+
+        # Perform in-place sorting on shared memory
+        wp.tile_sort(keys, values)
+
+        # Store sorted data back to global memory
+        wp.tile_store(sap_projection_lower, keys, offset=(base_idx,))
+        wp.tile_store(sap_sort_index, values, offset=(base_idx,))
+
+    return tile_sort_kernel
+
```
[the remark this answers is not in the document]
Comment (nvtw) [reply newton/_src/geometry/broad_phase_sap.py:134]: Your assumption is not correct: It is not required that the tile has the same size as the number of threads in the block

File: newton/_src/geometry/broad_phase_nxn.py [hunk 2]
```diff
@@ -180,31 +308,32 @@ def launch(
...
+                    geom_collision_group,
+                    geom_shape_world,
+                    self.world_cumsum_lower_tri,
+                    self.world_slice_ends,
+                    self.world_index_map,
+                    self.num_regular_worlds,
+                ],
+                outputs=[candidate_pair, num_candidate_pair, max_candidate_pair],
+                device=device,
+            )
```
[the remark this answers is not in the document]
Comment (nvtw) [reply newton/_src/geometry/broad_phase_nxn.py]: Won't do, it's the user's responsibility to not mix the devices

File: newton/tests/test_broad_phase.py [hunk 3]
```diff
@@ -19,10 +19,61 @@
...
+    Returns:
+        bool: True if the entities should collide, False otherwise
+    """
+    # Check world indices first
+    if world_a != -1 and world_b != -1 and world_a != world_b:
+        return False
+
+    # If same world or at least one is global (-1), check collision groups
+    return test_group_pair(collision_group_a, collision_group_b)
+
```
[the remark this answers is not in the document]
Comment (nvtw) [reply newton/tests/test_broad_phase.py:83]: I need to keep that code. Importing it from broad_phase_common leads to a huge performance degradation

File: newton/_src/geometry/__init__.py [hunk 4]
```diff
@@ -15,7 +15,7 @@
 
 from .broad_phase_common import test_group_pair, test_world_and_group_pair
 from .broad_phase_nxn import BroadPhaseAllPairs, BroadPhaseExplicit
-from .broad_phase_sap import BroadPhaseSAP
+from .broad_phase_sap import BroadPhaseSAP, SAPSortType
```
Comment (adenzler-nvidia) [inline newton/_src/geometry/__init__.py:18]: I'm not sure what our convention exactly is, but SAPSortType is public API so wouldn't it make sense to import from the public API.

File: newton/_src/geometry/broad_phase_common.py [hunk 5]
```diff
@@ -15,8 +15,11 @@
 
 from typing import Any
 
+import numpy as np
 import warp as wp
 
+from .flags import ShapeFlags
+
```
Comment (adenzler-nvidia) [inline newton/_src/geometry/broad_phase_common.py:22]: ShapeFlags can also be imported from the public API.

File: newton/_src/geometry/broad_phase_nxn.py [hunk 6]
```diff
@@ -53,11 +54,11 @@ def _nxn_broadphase_precomputed_pairs(
 
 
 @wp.func
-def _get_lower_triangular_indices(index: int, matrix_size: int) -> wp.vec2i:
+def _get_lower_triangular_indices(index: int, matrix_size: int):
```
Comment (adenzler-nvidia) [inline newton/_src/geometry/broad_phase_nxn.py]: you could still have an output type hint, or does it still not work with multiple return values?

File: newton/_src/geometry/broad_phase_nxn.py [hunk 7]
```diff
@@ -71,42 +72,107 @@ def _get_lower_triangular_indices(index: int, matrix_size: int) -> wp.vec2i:
...
+            world_id = mid
+        else:
+            low = mid + 1
+
+    # Calculate local index within this world
+    local_id = tid
+    if world_id > 0:
+        local_id = tid - world_cumsum_lower_tri[world_id - 1]
+
+    return world_id, local_id
```
Comment (adenzler-nvidia) [inline newton/_src/geometry/broad_phase_nxn.py:113]: Wondering if we should have special case for homogeneous worlds here that simplifies this indexing.

Can be a follow-up if you think this could be useful.

File: newton/_src/geometry/broad_phase_nxn.py [hunk 8]
```diff
@@ -71,42 +72,107 @@ def _get_lower_triangular_indices(index: int, matrix_size: int) -> wp.vec2i:
...
     world1 = shape_world[geom1]
     world2 = shape_world[geom2]
     collision_group1 = collision_group[geom1]
     collision_group2 = collision_group[geom2]
 
+    # Skip pairs where both geometries are global (world -1), unless we're in the dedicated -1 segment
+    # The dedicated -1 segment is the last segment (world_id >= num_regular_worlds)
+    is_dedicated_minus_one_segment = world_id >= num_regular_worlds
+    if world1 == -1 and world2 == -1 and not is_dedicated_minus_one_segment:
+        return
```
Comment (adenzler-nvidia) [inline newton/_src/geometry/broad_phase_nxn.py:169]: Would there be any benefit in actually also reordering the data for better coalescing? Most of this state is internal anyway, like the AABBs?

File: newton/_src/geometry/broad_phase_nxn.py [hunk 9]
```diff
@@ -180,31 +310,32 @@ def launch(
...
-                geom_count,
-                geom_cutoffs,
-                geom_collision_group,
-                geom_shape_world,
-            ],
-            outputs=[candidate_pair, num_candidate_pair, max_candidate_pair],
-            device=device,
-        )
+        # Launch with the precomputed number of kernel threads
+        if self.num_kernel_threads > 0:
```
Comment (adenzler-nvidia) [inline newton/_src/geometry/broad_phase_nxn.py]: can you check whether this actually results in a launch if dim is 0?

File: newton/_src/geometry/broad_phase_nxn.py [hunk 10]
```diff
@@ -257,6 +388,9 @@ def launch(
 
         num_candidate_pair.zero_()
 
+        if device is None:
+            device = geom_lower.device
+
```
Comment (adenzler-nvidia) [inline newton/_src/geometry/broad_phase_nxn.py:392]: was this a problem? It looks random, we're not doing this explicitly in many other places either..

File: newton/tests/test_broad_phase.py [hunk 11]
```diff
@@ -19,10 +19,61 @@
 import numpy as np
 import warp as wp
 
-from newton._src.geometry.broad_phase_common import test_group_pair
+from newton._src.geometry.broad_phase_sap import SAPSortType
+from newton._src.geometry.flags import ShapeFlags
 from newton.geometry import BroadPhaseAllPairs, BroadPhaseExplicit, BroadPhaseSAP
 
 
+def test_group_pair(group_a: int, group_b: int) -> bool:
```
Comment (adenzler-nvidia) [inline newton/tests/test_broad_phase.py:35]: can you follow coderabbit's suggestion and add a comment that explains why you need to copy this code?

File: newton/_src/geometry/__init__.py (hunk 4 above)
Comment (nvtw) [reply newton/_src/geometry/__init__.py:18]: Cursor tells me that files below _src should not import through the public API to avoid circular imports

File: newton/_src/geometry/broad_phase_common.py (hunk 5 above)
Comment (nvtw) [reply newton/_src/geometry/broad_phase_common.py:22]: Same as above

File: newton/_src/geometry/broad_phase_nxn.py (hunk 7 above)
Comment (nvtw) [reply newton/_src/geometry/broad_phase_nxn.py:113]: We probably would need to measure but I don't think the current implementation has much overhead due to the support of inhomogeneous worlds.

File: newton/_src/geometry/broad_phase_nxn.py (hunk 8 above)
Comment (nvtw) [reply newton/_src/geometry/broad_phase_nxn.py:169]: I assume there would be a benefit at the cost of a sort. The question is does the better coalescing amortize the cost of the sort, we would need to measure as well.

File: newton/_src/geometry/broad_phase_nxn.py (hunk 10 above)
Comment (nvtw) [reply newton/_src/geometry/broad_phase_nxn.py:392]: I think CodeRabbit wanted to have something like that, but not sure anymore. I can remove it again if you like.

File: newton/tests/test_broad_phase.py (hunk 11 above)
Comment (nvtw) [reply newton/tests/test_broad_phase.py:35]: Done

File: newton/_src/geometry/broad_phase_nxn.py (hunk 9 above)
Comment (nvtw) [reply newton/_src/geometry/broad_phase_nxn.py]: I removed the if. Looks like warp will skip the launch internally for 0 thrads.

File: newton/_src/geometry/__init__.py (hunk 4 above)
Comment (adenzler-nvidia) [reply newton/_src/geometry/__init__.py:18]: that makes sense, I didn't realize. Forget about this then!

File: newton/_src/geometry/broad_phase_nxn.py (hunk 8 above)
Comment (adenzler-nvidia) [reply newton/_src/geometry/broad_phase_nxn.py:169]: let's keep this in mind, shouldn't block this PR

File: newton/_src/geometry/broad_phase_nxn.py (hunk 7 above)
Comment (adenzler-nvidia) [reply newton/_src/geometry/broad_phase_nxn.py:113]: ack, we can always follow up if this is an issue.
